File: scala/src/main/scala/part1/chapter5/LazyList.scala
package part1.chapter5

enum LazyList[+A]:
    case Empty
    case Cons(h: () => A, t: () => LazyList[A])
        def headOption: Option[A] =
            this match
                case Empty => None
                case Cons(h, _) => Some(h())

    def toList: List[A] =
        @annotation.tailrec
        def go(ll: LazyList[A], acc: List[A]): List[A] =
            ll match
                case Cons(h, t) => go(t(), h() :: acc)
                case Empty => acc.reverse
        go(this, Nil)

object LazyList:
    def cons[A](
        hd: => A, tl: => LazyList[A]
    ): LazyList[A] =
        lazy val head = hd
        lazy val tail = tl
        Cons(() => head, () => tail)

    def empty[A]: LazyList[A] = Empty

    def apply[A](as: A*): LazyList[A] =
        if as.isEmpty
        then empty
        else
            lazy val head = as.head
            lazy val tail = as.tail
            Cons(() => head, () => apply(tail*))

File: haskell/lib/FpInScala/Part1/Chapter5/LazyList.hs
module FpInScala.Part1.Chapter5.LazyList
  ( LazyList(Empty, Cons)
  , lazyList
  , headOption
  , toList
  ) where

data LazyList a = Empty | Cons a (LazyList a)
  deriving (Eq, Show)

lazyList :: [a] -> LazyList a
lazyList [] = Empty
lazyList (x:xs) = Cons x (lazyList xs)

headOption :: LazyList a -> Maybe a
headOption Empty = Nothing
headOption (Cons x xs) = Just x

toList :: LazyList a -> [a]
toList ll = go ll []
  where
    go :: LazyList a -> [a] -> [a]
    go Empty acc = reverse acc
    go (Cons h t) acc = go t (h : acc)
